src/endpoints/ws.ts

```diff
@@ -220,6 +354,54 @@ export default class WsEndpoint {
       .get('/', res => {
         res.end(`{"players":${this.app.storage.playerList.size}}`);
       })
+      .get('/actions', res => {
+        res.writeHeader('Content-type', 'application/json');
+        res.end(`[${this.moderatorActions.join(',\n')}]`);
+      })
+      .post('/actions', res => {
+        readRequest(
+          res,
+          requestData => {
+            this.onActionsPost(res, requestData);
+          },
+          () => {
+            this.log.error('failed to parse /actions POST');
+          }
+        );
+      })
+      .get('/players', res => {
+        const list = [];
+        const players = Array.from(this.app.storage.playerList.values());
+
+        for (let i = 0; i < players.length; i += 1) {
+          const player = players[i];
+
+          list.push({
+            name: player.name.current,
+            id: player.id.current,
+            captures: player.captures.current,
+            spectate: player.spectate.current,
+            kills: player.kills.current,
+            deaths: player.deaths.current,
+            score: player.score.current,
+            lastMove: player.times.lastMove,
+            ping: player.ping.current,
+            flag: player.flag.current,
+          });
+        }
+
+        res.writeHeader('Content-type', 'application/json');
+        res.end(JSON.stringify(list, null, 2));
+      })
+      .get('/admin', async function onAdminGet(res) {
+        res.onAborted(() => {});
+
+        try {
+          res.end(await readFile(app.config.admin.htmlPath));
+        } catch (e) {
+          res.end(`internal error: ${e}`);
+        }
```

Comment:
[An example](https://github.com/uNetworking/uWebSockets.js/blob/master/examples/AsyncFunction.js) from the lib author how to write the same without try/catch.

Reply:
The catch() is for the IO done by readFile(). Does that make sense?

Reply:
Ah, ok. I thought we could cache this file in memory so we don't have to read it every time, but we can keep it so we can make changes without restart.

Reply:
The same for the password file.. I was tempted to use stat() and conditionally reload when changed, but that seems just as prone to causing problems if someone DoSes the endpoints, so a real fix for this is probably some kind of request throttling mechanism.

Meanwhile it's just a few syscalls and memory copies, the files are tiny and will always be sitting in OS cache